pop-api/examples/dao/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+		// inactive balances from bloating the blockchain state and slowing down the network.
+		#[ink(constructor, payable)]
+		pub fn new(
+			token_id: TokenId,
+			voting_period: BlockNumber,
+			min_balance: Balance,
+		) -> Result<Self, Psp22Error> {
+			let instance = Self {
+				proposals: Mapping::default(),
+				members: Mapping::default(),
+				last_votes: Mapping::default(),
+				voting_period,
+				token_id,
+				proposals_created: 0,
+			};
+			let contract_id = instance.env().account_id();
+			api::create(token_id, contract_id, min_balance).map_err(Psp22Error::from)?;
+			instance.env().emit_event(Created {
+				id: token_id,
+				creator: contract_id,
+				admin: contract_id,
+			});
+
+			Ok(instance)
+		}
+
+		/// Allows members to create new spending proposals
+		///
+		/// # Parameters
+		/// - `beneficiary` - The account that will receive the payment
+		/// if the proposal is accepted.
+		/// - `amount` - Amount requested for this proposal
+		/// - `description` - Description of the proposal
+		#[ink(message)]
+		pub fn create_proposal(
+			&mut self,
+			beneficiary: AccountId,
+			amount: Balance,
+			description: Vec<u8>,
+		) -> Result<(), Error> {
+			let caller = self.env().caller();
+			let contract = self.env().account_id();
+			let current_block = self.env().block_number();
+			let proposal_id: u32 = self.proposals_created.saturating_add(1);
+			let vote_end =
+				current_block.checked_add(self.voting_period).ok_or(Error::ArithmeticOverflow)?;
+			if description.len() >= STRINGLIMIT.into() {
+				return Err(Error::StringLimitReached);
+			}
+			let proposal = Proposal {
+				description,
+				vote_start: current_block,
+				vote_end,
+				yes_votes: 0,
+				no_votes: 0,
+				executed: false,
+				beneficiary,
+				amount,
+				proposal_id,
+			};
+
+			self.proposals.insert(proposal_id, &proposal);
+
+			self.env()
+				.emit_event(Created { id: proposal_id, creator: caller, admin: contract });
+
+			Ok(())
+		}
+
+		/// Allows Dao's members to vote for a proposal
+		///
+		/// # Parameters
+		/// - `proposal_id` - Identifier of the proposal
+		/// - `approve` - Indicates whether the vote is in favor (true) or against (false) the
+		///   proposal.
+		#[ink(message)]
+		pub fn vote(&mut self, proposal_id: u32, approve: bool) -> Result<(), Error> {
+			let caller = self.env().caller();
+			let current_block = self.env().block_number();
+			let now = self.env().block_number();
+			// let props = self.proposals.clone();
+
+			let mut proposal = self.proposals.get(proposal_id).ok_or(Error::ProposalNotFound)?;
+
+			if current_block < proposal.vote_start || current_block > proposal.vote_end {
+				return Err(Error::VotingPeriodEnded);
+			}
+
+			let member = self.members.get(caller).ok_or(Error::NotAMember)?;
+
+			if member.last_vote >= proposal.vote_start {
+				return Err(Error::AlreadyVoted);
+			}
+
+			if approve {
+				proposal.yes_votes = proposal.yes_votes.saturating_add(member.voting_power);
```

Comment:
No action needed. This could be improved by refactoring a bit: 

```rs
let mute votes = match approve {
  true => proposal.yes_votes,
  false => proposal.no_votes
};

votes.saturating_add(member.voting_power);

```